### PR TITLE
*: handle dirty table in StmtTxn

### PR DESCRIPTION
**Pull request by tiancaiamao**
Changes based on this one https://github.com/pingcap/tidb/pull/5699, so I tagged it with [WIP], I'll rebase it after the former merge.

### CODE REVIEW

**Comment by shenli** [comment]
Please resolve the conflicts.

**Comment by tiancaiamao** [comment]
Done. @shenli 
1. DirtyTable, DirtyDB is exposed from executor package.
2. Write operation will not write to dirty table directly. 
    Instead, it log the operation, and `StmtCommit` will flush operations to dirty table.

This is the last fix, next I'll do some clean up @coocood ,
and also add a `failed` mark to StmtTxn, handle `StmtCommit` error (which should never happen) 
by rollback the whole transaction. @jackysp @wentaoxu

**Comment by tiancaiamao** [comment]
PTAL @coocood @jackysp @shenli

File: session.go [hunk 1]
```diff
@@ -916,9 +916,16 @@ func (s *session) DropPreparedStmt(stmtID uint32) error {
...
 
-	buf       kv.MemBuffer
-	mutations map[int64]*binlog.TableMutation
-	// TODO: DirtyDB need to be statement level too.
+	buf          kv.MemBuffer
+	mutations    map[int64]*binlog.TableMutation
+	dirtyTableOP []dirtyTableOperation
+}
+
+type dirtyTableOperation struct {
```
Comment (jackysp) [inline session.go]: Please add comments for this.

**Comment by jackysp** [comment]
Try this below:

```
mysql> show create table t;
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                     |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `i` int(11) NOT NULL,
  `j` int(11) DEFAULT NULL,
  PRIMARY KEY (`i`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

start two transactions in session 1 and session 2 at meantime, then

session 1:

```
mysql> select * from t;
+---+------+
| i | j    |
+---+------+
| 4 |    2 |
| 6 |    1 |
+---+------+
2 rows in set (0.00 sec)

mysql> update ignore t set i = 8;
Query OK, 2 rows affected, 1 warning (0.01 sec)

mysql> select * from t;
+---+------+
| i | j    |
+---+------+
| 6 |    1 |
| 8 |    2 |
+---+------+
2 rows in set (0.01 sec)

mysql> commit;
Query OK, 0 rows affected (0.01 sec)

mysql> select * from t;
+---+------+
| i | j    |
+---+------+
| 7 |    2 |
| 8 |    1 |
+---+------+
2 rows in set (0.00 sec)
```

session 2:

```
mysql> select * from t;
+---+------+
| i | j    |
+---+------+
| 4 |    2 |
| 6 |    1 |
+---+------+
2 rows in set (0.00 sec)

mysql> update ignore t set i = 7;
Query OK, 2 rows affected, 1 warning (0.00 sec)

mysql> select * from t;
+---+------+
| i | j    |
+---+------+
| 6 |    1 |
| 7 |    2 |
+---+------+
2 rows in set (0.01 sec)

mysql> commit;
Query OK, 0 rows affected (0.00 sec)
```

both OK, but in master branch, one of the two transactions will rollback.

File: context/context.go [hunk 2]
```diff
@@ -78,6 +79,8 @@ type Context interface {
 	StmtRollback()
 	// StmtGetMutation gets the binlog mutation for current statement.
 	StmtGetMutation(int64) *binlog.TableMutation
+	// StmtDirtyTablesOP logs the dirty tables operations for current statement.
```
Comment (lamxTyler) [inline context/context.go]: `StmtDirtyTablesOP` -> `StmtDirtyTables`.

File: executor/write.go [hunk 3]
```diff
@@ -448,12 +447,21 @@ func (e *DeleteExec) removeRowsInTblRowMap(tblRowMap tableRowMapType) error {
 	return nil
 }
 
+const (
+	// DirtyTableAddRow is the constants for dirty table operation type.
```
Comment (lamxTyler) [inline executor/write.go]: `constants` -> `constant`.

File: context/context.go (hunk 2 above)
Comment (tiancaiamao) [reply context/context.go]: I've change StmtDirtyTables to StmtDirtyTablesOP, because it means an operation.

File: context/context.go [hunk 4]
```diff
@@ -78,6 +79,8 @@ type Context interface {
...
+	StmtLogDirtyOP(op int, tid int64, handle int64, row []types.Datum)
```
Comment (coocood) [inline context/context.go]: `Log` can be misunderstood as write log message to file.
How about `StmtAddDirtyTableOp`?

Comment (tiancaiamao) [reply context/context.go]: Done. @coocood

**Comment by coocood** [comment]
LGTM

**Comment by wentaoxu** [comment]
LGTM

**Pull request merged**